src/object/cli_shard.c

```diff
@@ -1999,95 +2039,50 @@ obj_shard_query_recx_post(struct obj_query_key_cb_args *cb_args, uint32_t shard,
 	daos_recx_t		*tmp_recx;
 	daos_recx_t		 recx[2] = {0};
 	uint64_t		 end[2], tmp_end;
-	uint32_t		 tgt_idx;
-	bool			 parity_checked = false;
+	uint32_t		 tgt_off;
 	bool			 from_data_tgt;
 	struct daos_oclass_attr	*oca;
-	uint64_t		 stripe_rec_nr, cell_rec_nr, rx_idx;
+	uint64_t		dkey_hash;
+	uint64_t		 stripe_rec_nr, cell_rec_nr;
 
 	oca = obj_get_oca(cb_args->obj);
 	if (oca == NULL || !daos_oclass_is_ec(oca)) {
 		*result_recx = *reply_recx;
 		return;
 	}
 
-	tgt_idx = shard % obj_get_grp_size(cb_args->obj);
-	from_data_tgt = tgt_idx < obj_ec_data_tgt_nr(oca);
+	dkey_hash = obj_ec_dkey_hash_get(cb_args->obj, cb_args->dkey_hash);
+	from_data_tgt = is_ec_data_shard(shard, dkey_hash, oca);
+	tgt_off = obj_ec_shard_off(dkey_hash, oca, shard);
 	stripe_rec_nr = obj_ec_stripe_rec_nr(oca);
 	cell_rec_nr = obj_ec_cell_rec_nr(oca);
 	tmp_recx = &recx[0];
-re_check:
-	if (reply_recx->rx_idx & PARITY_INDICATOR) {
-		daos_recx_t *punched_recx = &okqo->okqo_recx_punched;
-
-		D_ASSERT(!from_data_tgt);
-		rx_idx = (reply_recx->rx_idx & (~PARITY_INDICATOR));
-		D_ASSERTF(rx_idx % cell_rec_nr == 0, "rx_idx "DF_X64
-			  ", cell_rec_nr "DF_U64"\n",
-			  rx_idx, cell_rec_nr);
-		D_ASSERTF(reply_recx->rx_nr % cell_rec_nr == 0,
-			  "rx_nr "DF_U64", cell_rec_nr "DF_U64"\n",
-			  reply_recx->rx_nr, cell_rec_nr);
-		rx_idx = obj_ec_idx_vos2daos(rx_idx, stripe_rec_nr, cell_rec_nr,
-					     0);
-		tmp_recx->rx_idx = rx_idx;
-		tmp_recx->rx_nr = stripe_rec_nr *
-				     (reply_recx->rx_nr / cell_rec_nr);
-
-		if (DAOS_RECX_END(*punched_recx) > 0) {
-			D_DEBUG(DB_IO, "shard %d punched extent "DF_U64" "DF_U64"\n",
-				shard, punched_recx->rx_idx, punched_recx->rx_nr);
-			D_ASSERT(DAOS_RECX_END(*punched_recx) >= tmp_recx->rx_idx);
-			if (DAOS_RECX_END(*punched_recx) < DAOS_RECX_END(*tmp_recx)) {
-				uint64_t r_end = DAOS_RECX_END(*tmp_recx);
-
-				tmp_recx->rx_idx = DAOS_RECX_END(*punched_recx);
-				tmp_recx->rx_nr = r_end - tmp_recx->rx_idx;
-			} else if (punched_recx->rx_idx > tmp_recx->rx_idx) {
-				tmp_recx->rx_nr = min(punched_recx->rx_idx,
-						      DAOS_RECX_END(*tmp_recx)) - tmp_recx->rx_idx;
-			} else {
-				tmp_recx->rx_nr = 0;
-				tmp_recx->rx_idx = 0;
-			}
-		}
-		D_DEBUG(DB_IO, "shard %d get recx "DF_U64" "DF_U64"\n",
-			shard, tmp_recx->rx_idx, tmp_recx->rx_nr);
-	} else {
-		/* data ext from data shard needs to convert to daos ext,
-		 * replica ext from parity shard needs not to convert.
-		 */
-		*tmp_recx = *reply_recx;
-		tmp_end = DAOS_RECX_PTR_END(tmp_recx);
-		if (from_data_tgt && tmp_end > 0) {
-			if (get_max) {
-				tmp_recx->rx_idx = max(tmp_recx->rx_idx,
-					rounddown(tmp_end - 1, cell_rec_nr));
-				tmp_recx->rx_nr = tmp_end - tmp_recx->rx_idx;
-			} else {
-				tmp_recx->rx_nr = min(tmp_end,
-					roundup(tmp_recx->rx_idx + 1,
-						cell_rec_nr)) -
-					tmp_recx->rx_idx;
-			}
```

Comment:
this cleanup only work if server-side with PR7845, but that patch is not available in 2.0 server (available in 2.2).
so if 2.4 client need to work with 2.0 server, probably should keep the original client-side code?

Reply:
I think due to ec parity rotation, 2.4 client will not work with 2.0 server, otherwise let me get it back in the next patch.